R/data_color.R

```diff
@@ -391,131 +393,118 @@ adjust_luminance <- function(colors,
+  colors <- grDevices::col2rgb(bgnd_color)
+  score <- colSums(colors * c(299, 587, 144)) / 1000
+  ifelse(score >= yiq_contrasted_threshold, dark, light)
+}
 
-      } else {
-        NA_character_
-      }
-    })
+#' Convert a single color in the rgba() string format to hexadecimal (#RRGGBBAA)
+#'
+#' This function will accept a single color in the rgba() string format (e.g.,
+#' "`rgba(255,170,0,0.5)`") and output a color in the hexadecimal format with an
+#' alpha component (#RRGGBBAA).
+#'
+#' @noRd
+rgba_to_hex <- function(color) {
+
+  channels <-
+    color %>%
+    tidy_gsub(pattern = "(rgba\\(|\\))", "") %>%
+    strsplit(",") %>%
+    unlist() %>%
+    as.numeric()
+
+  grDevices::rgb(
+    red = channels[1] / 255,
+    green = channels[2] / 255,
+    blue = channels[3] / 255,
+    alpha = channels[4]
+  )
 }
 
-#' With any input color, transform to hexadecimal with alpha
+#' With a vector of input colors return normalized color strings
+#'
+#' Input colors can be color names (e.g., `"green"`, `"steelblue"`, etc.),
+#' colors in hexadecimal format with or without an alpha component (either
+#' #RRGGBB or #RRGGBBAA), or color strings in the rgba() string format
+#' (e.g., "`rgba(255,170,0,0.5)`"). Output is the same length vector as the
+#' input but it will contain a mixture of either #RRGGBB colors (if the input
+#' alpha value for a color is 1) or rgba() string format colors (if the input
+#' alpha value for a color is not 1).
 #'
 #' @noRd
-normalize_color <- function(colors,
-                            alpha) {
+process_color <- function(colors, alpha = NULL) {
 
-  if (length(colors) == 0) {
-    stop("One or more colors must be provided", call. = FALSE)
+  col_matrix_to_rgba <- function(color_matrix) {
+
+    color_matrix %>%
+      as.data.frame() %>%
+      dplyr::mutate(
+        rgba = paste0("rgba(", r, ",", g, ",", b, ",", round(alpha, 2), ")")
+      ) %>%
+      dplyr::pull(rgba)
   }
 
-  if (!(length(colors) == length(alpha) || length(alpha) == 1)) {
-    stop("The length of `alpha` must be the same length as `colors` or `1`",
-         call. = FALSE)
+  color_matrix <-
+    vapply(
+      colors,
+      FUN.VALUE = character(1),
+      USE.NAMES = FALSE,
+      FUN = function(color) {
+        if (is_rgba_col(color)) color <- rgba_to_hex(color = color)
+        color
+      }) %>%
+    farver::decode_colour(alpha = TRUE)
+
+  if (!is.null(alpha)) {
+    color_matrix[, 4] <- alpha
   }
 
-  # Combine hexadecimal color with corresponding alpha
-  vapply(
-    colors,
-    FUN.VALUE = character(1),
-    USE.NAMES = FALSE,
-    function(color) {
-      if (!is_hex_col_w_alpha(color)) {
-        color <- paste0(get_hexcolor_vec(color), frac_alpha_to_hex(alpha))
-      }
+  colors <- rep(NA_character_, nrow(color_matrix))
 
-      color
-    }
-  )
-}
+  colors_no_alpha <- unname(color_matrix[, 4] == 1 & !is.na(color_matrix[, 1]))
+  colors_w_alpha  <- unname(color_matrix[, 4] != 1 & !is.na(color_matrix[, 1]))
```

Comment:
What would `is.na(color_matrix[,1])` mean? The whole color is `NA`?

Reply:
Yes. I removed those `!is.na(...)` stmts because we definitely *don't* want missing colors (i.e., `NA` inputs). I also added a stop statement to catch any incoming `NA`s.